File: app/lib/features/notifications/data/notifications_repository.dart
import 'package:sbc_contacts/core/network/api_client.dart';
import 'package:sbc_contacts/core/network/paginated.dart';
import 'package:sbc_contacts/features/notifications/domain/app_notification.dart';

class NotificationsRepository {
  NotificationsRepository(this._api);
  final ApiClient _api;

  Future<Paginated<AppNotification>> list({
    int page = 1,
    int limit = 30,
    bool unreadOnly = false,
  }) async {
    final data = await _api.get('/notifications', query: {
      'page': page,
      'limit': limit,
      if (unreadOnly) 'unreadOnly': true,
    }) as Map<String, dynamic>;
    return Paginated.fromJson(data, AppNotification.fromJson);
  }

  Future<int> unreadCount() async {
    final data = await _api.get('/notifications/unread-count') as Map<String, dynamic>;
    return (data['count'] as num?)?.toInt() ?? 0;
  }

  Future<void> markRead(String id) => _api.post('/notifications/$id/read');

  Future<void> markAllRead() => _api.post('/notifications/read-all');

  Future<void> registerDevice({required String platform, String? pushToken}) => _api.post(
        '/notifications/devices',
        body: {'platform': platform, if (pushToken != null) 'pushToken': pushToken},
      );
}

File: ios/SBCContacts/Features/Notifications/NotificationsRepository.swift
import Foundation

/// In-app notification (backend `Notification`).
struct AppNotification: Identifiable, Sendable, Equatable {
    var id: String
    var type: String
    var title: String
    var body: String
    var createdAt: Date
    var readAt: Date?

    var isRead: Bool { readAt != nil }
}

extension AppNotification: Decodable {
    private enum CodingKeys: String, CodingKey {
        case id, type, title, body, createdAt, readAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = c.string(.id) ?? ""
        type = c.string(.type) ?? ""
        title = c.string(.title) ?? ""
        body = c.string(.body) ?? ""
        createdAt = c.date(.createdAt) ?? Date(timeIntervalSince1970: 1_577_836_800)
        readAt = c.date(.readAt)
    }
}

struct NotificationsRepository: Sendable {
    let api: APIClient

    private struct Count: Decodable, Sendable {
        let count: Int

        private enum CodingKeys: String, CodingKey { case count }

        init(from decoder: Decoder) throws {
            count = (try decoder.container(keyedBy: CodingKeys.self)).int(.count) ?? 0
        }
    }

    private struct DeviceBody: Encodable, Sendable {
        let platform: String
        let pushToken: String?
    }

    func list(page: Int = 1, limit: Int = 30, unreadOnly: Bool = false) async throws -> Paginated<AppNotification> {
        var query: [URLQueryItem] = [
            .init(name: "page", value: String(page)),
            .init(name: "limit", value: String(limit)),
        ]
        if unreadOnly { query.append(.init(name: "unreadOnly", value: "true")) }
        return try await api.get("/notifications", query: query)
    }

    func unreadCount() async throws -> Int {
        let result: Count = try await api.get("/notifications/unread-count")
        return result.count
    }

    func markRead(id: String) async throws {
        try await api.post("/notifications/\(id)/read")
    }

    func markAllRead() async throws {
        try await api.post("/notifications/read-all")
    }

    func registerDevice(platform: String, pushToken: String?) async throws {
        try await api.post("/notifications/devices", body: DeviceBody(platform: platform, pushToken: pushToken))
    }
}
